lib/pbench/agent/tool_meister.py

```diff
@@ -1043,87 +1043,130 @@ def _send_client_status(self, status: str) -> int:
                 ret_val = 0
         return ret_val
 
-    def init_tools(self, data: dict[str, str]) -> int:
-        """init_tools - setup all registered tools which have data collectors.
-
-        The Tool Data Sink will be setting up the actual processes which
-        collect data from these tools.
+    def _create_tool_directory(self, directory: str) -> Path:
+        """Create a temporary tool directory suitable for persistent or
+        transient tool use.
 
         Arguments:
 
-            data: a dictionary of the arguments sent to the Tool Meister
+            directory:  the directory value provided by the "init" or "start"
+                        tools message.
+
+                        When the Tool Meister is run in a container the
+                        "directory" parameter will not map into its namespace,
+                        so we always consider containerized Tool Meisters as
+                        remote.
+
+        Returns a Path object for the created diretory.
+
+        Raises a ToolException if the given directory is local and can't be
+        resolved, or if a temporary directory can't be created.
 
-        Returns 0 on success, # of failures otherwise.
         """
-        # Name of the temporary tool data directory to use when invoking
-        # tools.  This is a local temporary directory when the Tool Meister is
-        # remote from the pbench controller.
-        if self._params.controller == self._params.hostname:
-            # This is the case when the Tool Meister instance is running on
-            # the same host as the controller.  We just use the directory
+        local_dir = Path(directory)
+        if self._params.controller == self._params.hostname and local_dir.exists():
+            # The Tool Meister instance is running on the same host as the
+            # controller (not in a container).  We just use the directory
             # given to us in the `start` message.
             try:
-                _dir = Path(data["directory"]).resolve(strict=True)
-            except Exception:
-                self.logger.exception(
-                    "Failed to access provided result directory, %s", data["directory"]
-                )
-                self._send_client_status("internal-error")
-                return 1
+                base_dir = local_dir.resolve(strict=True)
+            except Exception as exc:
+                raise ToolException(
+                    f"Failed to access tool directory, {directory}"
+                ) from exc
```

Comment:
I believe that this `from` expression is superfluous inside this `except` clause.

Reply:
Is [this](https://www.pythontutorial.net/python-oop/python-raise-from/) your understanding of what `raise from` does?

Reply:
Looking at the page you cited, I cannot find any difference between the two sample code blocks which are supposed to produce different behaviors for the `raise...from` structures.  😆

However, the examples [here](https://docs.python.org/3/reference/simple_stmts.html#the-raise-statement) show it:
- using a bare `raise` in an `except` block produces, "During handling of the above exception, another exception occurred"; while
- using `raise...from` in an `except` block yields, "The above exception was the direct cause of the following exception", which _is_ nicer, but I'm not sure it's worth the extra "stuff" in the code.

Reply:
The intention as coded is to leverage this feature.  Exactly what are you saying is extra that is not worth it?

It would seem the addition of "as exc" on line 1073 and "from exc" on line 1076 are unobtrusive.

Do these additions pose a substantive maintenance issue?

Reply:
From where I stand, we add a local variable and an additional reference to it (i.e., the `as` and `from` clauses), and what it accomplishes is a subtle change in one line in the middle of the outputted text, which I'm not sure that anyone will actually notice or read.

No, the additions don't pose a substantive maintenance issue, but the benefit they produce isn't particularly impressive, either.  I'm inclined toward the reduction in the source code...but, since you obviously think it's worthwhile, I'll seek satisfaction in the more explicit text, instead.